File: leet/0040-combination-sum-ii/combination-sum-ii.go
import (
    "sort"
)

var result [][]int = make([][]int, 0)
var stack []int = make([]int, 0)

func combinationSum2(candidates []int, target int) [][]int {
    result = make([][]int, 0)
    stack = make([]int, 0)
    if len(candidates) == 0 {return result}
    
    sort.Ints(candidates)
    check(candidates, target, 0)
    return result
}

func check(candidates []int, target int, start int){
    if target < 0 {return}
    if target == 0{
        r := make([]int, len(stack))
        for i, num := range stack {
            r[i] = num
        }
        result = append(result, r)
        return
    }
    
    for i := start; i < len(candidates); i++ {
        stack = append(stack, candidates[i])
        check(candidates, target - candidates[i], i + 1)
        stack = stack[:len(stack) - 1]
        
        for i + 1 < len(candidates) && candidates[i] == candidates[i + 1] {
            i++
        }
    }
}

File: leet/0040-combination-sum-ii/combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        if len(candidates) == 0 : return []
        
        stack = []
        ret = []
        candidates.sort()
        
        def dfs(s: int, t: int):
            
            if t == 0:
                ret.append([] + stack)
                return
            elif t < 0:
                return
            
            i = s
            while i < len(candidates) and t >= candidates[i]:
                stack.append(candidates[i])
                dfs(i + 1,  t - candidates[i])
                stack.pop()

                while i + 1 < len(candidates) and candidates[i] == candidates[i + 1]:
                    i += 1
                i += 1
                
        dfs(0, target)
        return ret
            
            
        
